rust/loqui_protocol/src/frames.rs

```diff
@@ -0,0 +1,530 @@
+use crate::error::ProtocolError;
+use byteorder::{BigEndian, ByteOrder};
+use bytes::{BufMut, BytesMut};
+use std::str::from_utf8;
+
+type DecodeResult<T> = Result<Option<T>, ProtocolError>;
+
+#[derive(Debug, PartialEq)]
+pub enum LoquiFrame {
+    Hello(Hello),
+    HelloAck(HelloAck),
+    Ping(Ping),
+    Pong(Pong),
+    Request(Request),
+    Response(Response),
+    Push(Push),
+    GoAway(GoAway),
+    Error(Error),
+}
+
+pub trait Frame: Sized + 'static {
+    ///
+    /// Opcode of the frame.
+    ///
+    const OPCODE: u8;
+    ///
+    /// Header size in bytes, including the payload size field.
+    ///
+    const HEADER_SIZE_IN_BYTES: usize;
+
+    ///
+    /// Put the header bytes into the destination buffer. Space is already reserved. Do not
+    /// put the payload size, that will be handled for you based on what is returned by
+    /// Frame::payload()
+    ///
+    fn put_header(&self, dst: &mut BytesMut);
+    ///
+    /// Return the payload bytes that should be encoded.
+    ///
+    fn payload(self) -> Option<Vec<u8>>;
+    ///
+    /// Read the payload size from the buffer.
+    ///
+    fn read_payload_size(buf: &mut BytesMut) -> u32;
+    ///
+    /// Given a buf that is a complete frame, parse and return the `Frame`.
+    ///
+    fn from_buf(buf: &BytesMut) -> DecodeResult<Self>;
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Hello {
+    pub flags: u8,
+    pub version: u8,
+    pub encodings: Vec<String>,
+    pub compressions: Vec<String>,
+}
+
+impl Frame for Hello {
+    const OPCODE: u8 = 1;
+    const HEADER_SIZE_IN_BYTES: usize = 7;
+
+    #[inline]
```

Comment:
Are you asking for this `Hello` frame in particular or all of the `inline`s in this file?

Frame binary encoding/decoding is a hot path. I assumed it would be best to inline it. Maybe I am wrong.

Reply:
I don't know if this applies to Rust, but conventional wisdom for C++ says you should avoid declaring something inline unless you notice that you *need* to -- the compiler will typically inline things for you if it can & it makes sense to do so.

I don't know how much this matters either way though. 🤷‍♂